libs/wai-utilities/src/Network/Wai/Utilities/Server.hs

```diff
@@ -184,21 +166,39 @@ measureRequests m rtree = withPathTemplate rtree $ \p ->
       requestCounter m p . duration 30 12 m p
 {-# INLINEABLE measureRequests #-}
 
+-- | Alias for @'catchErrors'' True@.
+catchErrors  :: Logger -> OnErrorMetrics -> Middleware
+catchErrors = catchErrors' True
+{-# INLINEABLE catchErrors #-}
+
 -- | Create a middleware that catches exceptions and turns
 -- them into appropriate 'Error' responses, thereby logging
 -- as well as counting server errors (i.e. exceptions that
 -- yield 5xx responses).
-catchErrors :: Logger -> Metrics -> Middleware
-catchErrors l m app req k =
-    app req k `catch` errorResponse
+--
+-- The boolean determines whether response values with
+-- status >= 400 are allowed (servant), or whether we should
+-- not expect and consequently warn about them (wai-route).
+catchErrors' :: Bool -> Logger -> OnErrorMetrics -> Middleware
+catchErrors' warnNonExceptFailures l m app req k =
+    app req k' `catch` errorResponse
   where
     errorResponse ex = do
         er <- runHandlers ex errorHandlers
         when (statusCode (Error.code er) >= 500) $
-            logIO l Log.Error (Just req) (oneline <$> show ex)
+            logIO l Log.Error (Just req) (show ex)
         onError l m req k er
-    oneline c = if isSpace c then ' ' else c
-{-# INLINEABLE catchErrors #-}
+
+    -- catch status >= 400 even if it does not come in the form of an exception, but as a
+    -- response, and log the request id.  not much else we can do about it without taking
+    -- apart the streaming response.  this should be fixed whenever it is found in the logs.
+    k' resp = do
```

Comment:
I know you probably didn't write this originally; but can we use something more descriptive than `k` please? I have no clue what it stands for or does :)

I'm personally in favour of type sigs on `where` clauses too; but not sure if we have a ruling on that yet.

Reply:
Noted and full-heartedly agreed.  I won't change anything in this PR, though.